contracts/abstracted_account.algo.ts

```diff
@@ -35,35 +42,169 @@ export class AbstractedAccount extends Contract {
    */
   namedPlugins = BoxMap<bytes, PluginsKey>({ prefix: 'n' });
 
+  private assertPluginCallAllowed(app: AppID, caller: Address): void {
+    const key: PluginsKey = { application: app, allowedCaller: caller };
+
+    assert(this.plugins(key).exists, 'plugin does not exist');
+    assert(this.plugins(key).value.lastValidRound >= globals.round, 'plugin is expired');
+    assert(
+      globals.round - this.plugins(key).value.lastCalled >= this.plugins(key).value.cooldown,
+      'plugin is on cooldown'
+    );
+  }
+
+  private pluginCallAllowed(app: AppID, caller: Address): boolean {
+    const key: PluginsKey = { application: app, allowedCaller: caller };
+
+    return (
+      this.plugins(key).exists &&
+      this.plugins(key).value.lastValidRound >= globals.round &&
+      globals.round - this.plugins(key).value.lastCalled >= this.plugins(key).value.cooldown
+    );
+  }
+
+  private ensuresRekeyBack(txn: Txn): boolean {
```

Comment:
Nit, but I think a name like `txnRekeysBack` is a more descriptive name for this function. `ensures` just feels like the wrong word here but might just be me.